Modules/_io/fileio.c

```diff
@@ -267,9 +272,12 @@ _io_FileIO___init___impl(fileio *self, PyObject *nameobj, const char *mode,
 #endif
     if (self->fd >= 0) {
         if (self->closefd) {
-            /* Have to close the existing file first. */
-            if (internal_close(self) < 0)
+            /* Have to close the existing file first.
+
+               This leaves the stat so we can reuse the memory. */
```

Comment:
I don't think that this optimization is worth it, I would prefer to always clear the cache in close().